File: src/frontend/src/api/finance/closing.ts
/**
 * M4 재무/회계 — 결산/재무제표 API 호출 함수
 */
import api from '../client';

// ── 타입 정의 ──

export interface TrialBalanceRow {
  account_id: string;
  account_code: string;
  account_name: string;
  account_type: string;
  total_debit: number;
  total_credit: number;
  balance: number;
}

export interface TrialBalanceResponse {
  rows: TrialBalanceRow[];
  total_debit: number;
  total_credit: number;
  start_date: string | null;
  end_date: string | null;
}

export interface IncomeStatementRow {
  account_id: string;
  account_code: string;
  account_name: string;
  amount: number;
}

export interface IncomeStatementResponse {
  revenue_items: IncomeStatementRow[];
  expense_items: IncomeStatementRow[];
  total_revenue: number;
  total_expense: number;
  net_income: number;
  start_date: string | null;
  end_date: string | null;
}

export interface BalanceSheetRow {
  account_id: string;
  account_code: string;
  account_name: string;
  amount: number;
}

export interface BalanceSheetResponse {
  asset_items: BalanceSheetRow[];
  liability_items: BalanceSheetRow[];
  equity_items: BalanceSheetRow[];
  total_assets: number;
  total_liabilities: number;
  total_equity: number;
  net_income: number;
  as_of_date: string | null;
}

export interface ClosingParams {
  start_date?: string;
  end_date?: string;
  as_of_date?: string;
  fiscal_year_id?: string;
}

// ── API 함수 ──

export async function fetchTrialBalance(params: ClosingParams = {}) {
  const res = await api.get('/finance/closing/trial-balance', { params });
  return res.data.data as TrialBalanceResponse;
}

export async function fetchIncomeStatement(params: ClosingParams = {}) {
  const res = await api.get('/finance/closing/income-statement', { params });
  return res.data.data as IncomeStatementResponse;
}

export async function fetchBalanceSheet(params: ClosingParams = {}) {
  const res = await api.get('/finance/closing/balance-sheet', { params });
  return res.data.data as BalanceSheetResponse;
}

export async function closePeriod(fiscalYearId: string) {
  const res = await api.post('/finance/closing/close-period', {
    fiscal_year_id: fiscalYearId,
  });
  return res.data;
}

export async function reopenPeriod(fiscalYearId: string) {
  const res = await api.post('/finance/closing/reopen-period', {
    fiscal_year_id: fiscalYearId,
  });
  return res.data;
}

File: src/backend/modules/m4_finance/schemas/closing.py
"""
M4 재무/회계 — 결산/재무제표 Pydantic 스키마
"""
from datetime import date
from typing import Optional
from pydantic import BaseModel, Field


class TrialBalanceRow(BaseModel):
    """시산표 행"""
    account_id: str
    account_code: str
    account_name: str
    account_type: str
    total_debit: float
    total_credit: float
    balance: float  # 차변 합계 - 대변 합계 (자산/비용: 양수 정상, 부채/자본/수익: 음수 정상)


class TrialBalanceResponse(BaseModel):
    """시산표 응답"""
    rows: list[TrialBalanceRow]
    total_debit: float
    total_credit: float
    start_date: Optional[date] = None
    end_date: Optional[date] = None


class IncomeStatementRow(BaseModel):
    """손익계산서 항목"""
    account_id: str
    account_code: str
    account_name: str
    amount: float


class IncomeStatementResponse(BaseModel):
    """손익계산서 응답"""
    revenue_items: list[IncomeStatementRow]
    expense_items: list[IncomeStatementRow]
    total_revenue: float
    total_expense: float
    net_income: float  # 당기순이익
    start_date: Optional[date] = None
    end_date: Optional[date] = None


class BalanceSheetRow(BaseModel):
    """재무상태표 항목"""
    account_id: str
    account_code: str
    account_name: str
    amount: float


class BalanceSheetResponse(BaseModel):
    """재무상태표 응답"""
    asset_items: list[BalanceSheetRow]
    liability_items: list[BalanceSheetRow]
    equity_items: list[BalanceSheetRow]
    total_assets: float
    total_liabilities: float
    total_equity: float
    net_income: float  # 당기순이익
    as_of_date: Optional[date] = None


class ClosePeriodRequest(BaseModel):
    """기간 마감 요청"""
    fiscal_year_id: str = Field(..., description="마감할 회계연도 ID")
